bin/tools/utils.js

```diff
@@ -216,6 +218,379 @@ fs.copyFileSync("bin/build/qx.cmd", "bootstrap/qx.cmd");
+          return stat(to)
+            .then(tmp => statTo = tmp)
+            .catch(err => {
+              if (err.code !== "ENOENT") {
+                throw err;
+              }
+            });
+        })
+        .then(() => {
+          if (!statTo || statFrom.isDirectory() != statTo.isDirectory()) {
+            return t.deleteRecursive(to)
+              .then(() => copy(statFrom, statTo));
+          } else if (statFrom.isDirectory() || (statFrom.mtime.getTime() > statTo.mtime.getTime() || statFrom.size != statTo.size)) {
+            return copy(statFrom, statTo);
+          }
+          return undefined;
+        })
+        .then(resolve)
+        .catch(reject);
+    });
+  },
+
+  /**
+   * Copies a file
+   * @param from {String} path to copy from
+   * @param to {String} path to copy to
+   * @async
+   */
+  copyFile: function(from, to) {
+    return new Promise((resolve, reject) => {
+      moreUtils.mkParentPath(to, function() {
+        var rs = fs.createReadStream(from, { flags: "r", encoding: "binary" });
+        var ws = fs.createWriteStream(to, { flags: "w", encoding: "binary" });
+        rs.on("end", function() {
+          resolve(from, to);
+        });
+        rs.on("error", reject);
+        ws.on("error", reject);
+        rs.pipe(ws);
+      });
+    });
+  },
+
+  /**
+   * Returns the stats for a file, or null if the file does not exist
+   *
+   * @param filename
+   * @returns {fs.Stat}
+   * @async
+   */
+  safeStat: function(filename) {
+    return new Promise((resolve, reject) => {
+      fs.stat(filename, function(err, stats) {
+        if (err && err.code != "ENOENT") {
+          reject(err);
+        } else {
+          resolve(err ? null : stats);
+        }
+      });
+    });
+  },
+
+  /**
+   * Deletes a file, does nothing if the file does not exist
+   *
+   * @param filename {String} file to delete
+   * @async
+   */
+  safeUnlink: function(filename) {
+    return new Promise((resolve, reject) => {
+      fs.unlink(filename, function(err) {
+        if (err && err.code != "ENOENT") {
+          reject(err);
+        } else {
+          resolve();
+        }
+      });
+    });
+  },
+
+  /**
+   * Renames a file, does nothing if the file does not exist
+   *
+   * @param from {String} file to rename
+   * @param to {String} new filename
+   * @async
+   */
+  safeRename: function(from, to) {
+    return new Promise((resolve, reject) => {
+      fs.rename(from, to, function(err) {
+        if (err && err.code != "ENOENT") {
+          reject(err);
+        } else {
+          resolve();
+        }
+      });
+    });
+  },
+
+  /**
+   * Rotates files so that this file does not exist, by renaming the existing file to have a ".1"
+   * appended, and the ".1" to be renamed to ".2" etc, up to `length` versions
+   *
+   * @param filename {String} filename to rotate
+   * @param length {Integer} maximum number of files
+   * @async
+   */
+  rotateUnique: async function(filename, length) {
```

Comment:
Oh, I just copy & pasted utility methods from  `qx.tool.utils.*` to make them available to the test methods, which do not have qx available. The state of utility methods in `qx.tool` is a real mess, which has to do with its ad-hoc and decentral development. There is a lot of redundancy and overlapping methods, plus duplicate functionality targeting earlier node versions. It needs a thorough clean-up, but that means a lot of work. In particular, we need some classes which can be accessed both by qx and the nodeJS runtime - which would probably work with the "qx.server" package.

The method you point to is probably not needed by the test pipeline and can be removed.

Reply:
I think `.gitkeep` is needed because git doesn't store empty directories. I need the directoy because it must be renamed by the migration.

Reply:
@hkollmann if it's not used anywhere then it can be deleted; I recognise it as one of mine, I use it for rotating log files, but it's not very well defined (or sophisticated) so does not need to be in the framework unless its used

Reply:
@cboulanger : Could you delete it? Inclusive the then not needed test?

Reply:
I will delete it.

Reply:
Done.